File: keyboards/ios/LisanUdDawat/LSDLearningKB/KeyboardSettings.swift
import Foundation

enum KeyboardSettings {
    // Shared with the main app via the App Group container so settings changed
    // in either place are visible to both.  Falls back to standard if the group
    // entitlement is not provisioned (e.g. during unit tests).
    static let defaults = UserDefaults(suiteName: "group.com.exordiumnetworks.lsdkeyboard") ?? .standard

    static var predictionEnabled: Bool {
        get { defaults.bool(forKey: "prediction_enabled") }
        set { defaults.set(newValue, forKey: "prediction_enabled") }
    }

    // Experimental: on the Latin layer, transliterate the typed Roman word
    // into LSD candidates in the predictive bar.
    static var translitEnabled: Bool {
        get { defaults.bool(forKey: "translit_enabled") }
        set { defaults.set(newValue, forKey: "translit_enabled") }
    }

    // Pause corpus collection without clearing stored data.
    static var corpusEnabled: Bool {
        get { defaults.object(forKey: "corpus_enabled") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "corpus_enabled") }
    }

    // Double-tap a key to insert its secondary character.
    static var doubleTapEnabled: Bool {
        get { defaults.object(forKey: "double_tap_enabled") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "double_tap_enabled") }
    }

    enum DelayPreset: String {
        case short  = "short"   // 0.25s
        case normal = "normal"  // 0.35s
        case long   = "long"    // 0.50s
        case custom = "custom"
    }

    static var doubleTapDelayPreset: DelayPreset {
        get { DelayPreset(rawValue: defaults.string(forKey: "double_tap_delay_preset") ?? "normal") ?? .normal }
        set { defaults.set(newValue.rawValue, forKey: "double_tap_delay_preset") }
    }

    // Actual delay in seconds; presets write here, custom stepper writes here directly.
    static var doubleTapDelay: TimeInterval {
        get {
            let v = defaults.double(forKey: "double_tap_delay")
            return v > 0 ? v : 0.35
        }
        set { defaults.set(newValue, forKey: "double_tap_delay") }
    }

    // CRULP Urdu yeh: Farsi yeh ی (no dots, U+06CC) vs Arabic yeh ي (two dots, U+064A).
    enum UrduYehStyle: String {
        case farsiYeh  = "farsi_yeh"    // ی  default — standard in Urdu/Farsi
        case arabicYeh = "arabic_yeh"   // ي
    }

    static var urduYehStyle: UrduYehStyle {
        get { UrduYehStyle(rawValue: defaults.string(forKey: "urdu_yeh_style") ?? "") ?? .arabicYeh }
        set { defaults.set(newValue.rawValue, forKey: "urdu_yeh_style") }
    }

    // Double alef (اا) produces either اٰ (alef with kharo zabar) or آ (alef madda).
    enum DoubleAlefStyle: String {
        case kharoZabar = "kharo_zabar"   // اٰ  default
        case alefMadda  = "alef_madda"    // آ
    }

    static var doubleAlefStyle: DoubleAlefStyle {
        get { DoubleAlefStyle(rawValue: defaults.string(forKey: "double_alef_style") ?? "") ?? .kharoZabar }
        set { defaults.set(newValue.rawValue, forKey: "double_alef_style") }
    }

    // Kaaf style: Arabic ك (U+0643) vs Urdu/Farsi ک (U+06A9).
    enum KaafStyle: String {
        case arabic = "arabic_kaaf"   // ك  default
        case urdu   = "urdu_kaaf"     // ک
    }

    static var kaafStyle: KaafStyle {
        get { KaafStyle(rawValue: defaults.string(forKey: "kaaf_style") ?? "") ?? .arabic }
        set { defaults.set(newValue.rawValue, forKey: "kaaf_style") }
    }

    // Haa style: Arabic ه (U+0647) vs Urdu He Goal ہ (U+06C1).
    enum HaaStyle: String {
        case arabic = "arabic_haa"    // ه  default
        case urdu   = "urdu_haa"      // ہ
    }

    static var haaStyle: HaaStyle {
        get { HaaStyle(rawValue: defaults.string(forKey: "haa_style") ?? "") ?? .arabic }
        set { defaults.set(newValue.rawValue, forKey: "haa_style") }
    }

    // Taa marbuta style: Arabic ة (U+0629) vs Urdu ۃ (U+06C3).
    enum TaaMarbuta: String {
        case arabic = "arabic_taa"    // ة  default
        case urdu   = "urdu_taa"      // ۃ
    }

    static var taaMarbuta: TaaMarbuta {
        get { TaaMarbuta(rawValue: defaults.string(forKey: "taa_marbuta") ?? "") ?? .arabic }
        set { defaults.set(newValue.rawValue, forKey: "taa_marbuta") }
    }

    // Active keyboard layout.
    enum LayoutType: String {
        case lsd            = "lsd"
        case arabicStandard = "arabic_standard"
        case crulpUrdu      = "crulp_urdu"
    }

    static var selectedLayout: LayoutType {
        get { LayoutType(rawValue: defaults.string(forKey: "selected_layout") ?? "lsd") ?? .lsd }
        set { defaults.set(newValue.rawValue, forKey: "selected_layout") }
    }

    // Long-press delay before the alternate-characters popup appears.
    // Presented as Short (0.20s) / Normal (0.35s) / Long (0.50s) in settings.
    static var longPressDelay: TimeInterval {
        get {
            let v = defaults.double(forKey: "long_press_delay")
            return v > 0 ? v : 0.35
        }
        set { defaults.set(newValue, forKey: "long_press_delay") }
    }

    // Interval at which a held long-press alternate auto-repeats. 0 = disabled.
    // Presented as Off / Slow (0.25s) / Fast (0.10s) in settings.
    static var popupRepeatInterval: TimeInterval {
        get {
            guard defaults.object(forKey: "popup_repeat_interval") != nil else { return 0.1 }
            return defaults.double(forKey: "popup_repeat_interval")
        }
        set { defaults.set(newValue, forKey: "popup_repeat_interval") }
    }

    // One-time tooltip for the BiDi fix button.
    static var biDiTooltipShown: Bool {
        get { defaults.bool(forKey: "bidi_tooltip_shown") }
        set { defaults.set(newValue, forKey: "bidi_tooltip_shown") }
    }

    // One-time tooltip for the Latin (AaBb) key.
    static var latinKeyTooltipShown: Bool {
        get { defaults.bool(forKey: "latin_key_tooltip_shown") }
        set { defaults.set(newValue, forKey: "latin_key_tooltip_shown") }
    }

    // Depth-gradient bevel on keys to simulate a forward-tilted physical surface.
    // Top of each key face is slightly shadowed; intensity varies by row.
    static var angledKeysEnabled: Bool {
        get { defaults.object(forKey: "angled_keys_enabled") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "angled_keys_enabled") }
    }
}

File: keyboards/android/app/src/main/java/com/lisanuddawat/keyboard/KeyboardSettings.kt
package com.lisanuddawat.keyboard

import android.content.Context

object KeyboardSettings {

    private fun prefs(ctx: Context) =
        ctx.getSharedPreferences("lsd_keyboard_settings", Context.MODE_PRIVATE)

    // ── Layout ──────────────────────────────────────────────────────────

    enum class LayoutType(val key: String) {
        LSD("lsd"), ARABIC_STANDARD("arabic_standard"), CRULP_URDU("crulp_urdu")
    }

    fun getLayout(ctx: Context): LayoutType =
        LayoutType.values().firstOrNull { it.key == prefs(ctx).getString("selected_layout", "lsd") }
            ?: LayoutType.LSD

    fun setLayout(ctx: Context, v: LayoutType) =
        prefs(ctx).edit().putString("selected_layout", v.key).apply()

    // ── Predictions ──────────────────────────────────────────────────────

    fun getPredictions(ctx: Context): Boolean =
        prefs(ctx).getBoolean("prediction_enabled", false)

    fun setPredictions(ctx: Context, v: Boolean) =
        prefs(ctx).edit().putBoolean("prediction_enabled", v).apply()

    // ── Double-press ─────────────────────────────────────────────────────

    fun getDoublePressEnabled(ctx: Context): Boolean =
        prefs(ctx).getBoolean("double_press_enabled", true)

    fun setDoublePressEnabled(ctx: Context, v: Boolean) =
        prefs(ctx).edit().putBoolean("double_press_enabled", v).apply()

    fun getDoublePressWindowMs(ctx: Context): Long =
        prefs(ctx).getLong("double_press_window_ms", 400L)

    fun setDoublePressWindowMs(ctx: Context, v: Long) =
        prefs(ctx).edit().putLong("double_press_window_ms", v).apply()

    // ── Double-alef style ────────────────────────────────────────────────

    enum class DoubleAlefStyle(val key: String) {
        KHARO_ZABAR("kharo_zabar"),  // اٰ  default
        ALEF_MADDA("alef_madda")    // آ
    }

    fun getDoubleAlefStyle(ctx: Context): DoubleAlefStyle =
        DoubleAlefStyle.values().firstOrNull {
            it.key == prefs(ctx).getString("double_alef_style", "kharo_zabar")
        } ?: DoubleAlefStyle.KHARO_ZABAR

    fun setDoubleAlefStyle(ctx: Context, v: DoubleAlefStyle) =
        prefs(ctx).edit().putString("double_alef_style", v.key).apply()

    // ── Kaaf style ───────────────────────────────────────────────────────

    enum class KaafStyle(val key: String) {
        ARABIC_KAAF("arabic_kaaf"),   // ك  default
        URDU_KAAF("urdu_kaaf")        // ک
    }

    fun getKaafStyle(ctx: Context): KaafStyle =
        KaafStyle.values().firstOrNull { it.key == prefs(ctx).getString("kaaf_style", "arabic_kaaf") }
            ?: KaafStyle.ARABIC_KAAF

    fun setKaafStyle(ctx: Context, v: KaafStyle) =
        prefs(ctx).edit().putString("kaaf_style", v.key).apply()

    // ── Haa style ────────────────────────────────────────────────────────

    enum class HaaStyle(val key: String) {
        ARABIC_HAA("arabic_haa"),     // ه  default
        URDU_HAA("urdu_haa")          // ہ
    }

    fun getHaaStyle(ctx: Context): HaaStyle =
        HaaStyle.values().firstOrNull { it.key == prefs(ctx).getString("haa_style", "arabic_haa") }
            ?: HaaStyle.ARABIC_HAA

    fun setHaaStyle(ctx: Context, v: HaaStyle) =
        prefs(ctx).edit().putString("haa_style", v.key).apply()

    // ── Taa marbuta style ────────────────────────────────────────────────

    enum class TaaMarbuta(val key: String) {
        ARABIC_TAA("arabic_taa"),     // ة  default
        URDU_TAA("urdu_taa")          // ۃ
    }

    fun getTaaMarbuta(ctx: Context): TaaMarbuta =
        TaaMarbuta.values().firstOrNull { it.key == prefs(ctx).getString("taa_marbuta", "arabic_taa") }
            ?: TaaMarbuta.ARABIC_TAA

    fun setTaaMarbuta(ctx: Context, v: TaaMarbuta) =
        prefs(ctx).edit().putString("taa_marbuta", v.key).apply()

    // ── Yeh style (all layouts) ──────────────────────────────────────────

    enum class YehStyle(val key: String) {
        FARSI_YEH("farsi_yeh"),    // ی  default — standard in Urdu/Farsi
        ARABIC_YEH("arabic_yeh")  // ي
    }

    fun getYehStyle(ctx: Context): YehStyle =
        YehStyle.values().firstOrNull {
            it.key == prefs(ctx).getString("urdu_yeh_style", "arabic_yeh")
        } ?: YehStyle.ARABIC_YEH

    fun setYehStyle(ctx: Context, v: YehStyle) =
        prefs(ctx).edit().putString("urdu_yeh_style", v.key).apply()

    // ── BiDi tooltip ─────────────────────────────────────────────────────

    fun getBiDiTooltipShown(ctx: Context): Boolean =
        prefs(ctx).getBoolean("bidi_tooltip_shown", false)

    fun setBiDiTooltipShown(ctx: Context) =
        prefs(ctx).edit().putBoolean("bidi_tooltip_shown", true).apply()

    // ── Long-press behaviour ─────────────────────────────────────────────

    fun getLongPressDelayMs(ctx: Context): Long =
        prefs(ctx).getLong("long_press_delay_ms", 350L)

    fun setLongPressDelayMs(ctx: Context, v: Long) =
        prefs(ctx).edit().putLong("long_press_delay_ms", v).apply()

    // 0 = repeat disabled
    fun getPopupRepeatIntervalMs(ctx: Context): Long =
        prefs(ctx).getLong("popup_repeat_interval_ms", 100L)

    fun setPopupRepeatIntervalMs(ctx: Context, v: Long) =
        prefs(ctx).edit().putLong("popup_repeat_interval_ms", v).apply()

    // ── Latin key tooltip ─────────────────────────────────────────────────

    fun getLatinKeyTooltipShown(ctx: Context): Boolean =
        prefs(ctx).getBoolean("latin_key_tooltip_shown", false)

    fun setLatinKeyTooltipShown(ctx: Context) =
        prefs(ctx).edit().putBoolean("latin_key_tooltip_shown", true).apply()
}
